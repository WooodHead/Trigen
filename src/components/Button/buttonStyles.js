/**
 * Creates button specific styles which can be applied to different elements/components
 *
 * @module buttonStyles
  */

import { css } from 'styled-components';


const buttonStyles = css`
  display: inline-block;
  box-sizing: border-box;
  text-decoration: none;
  outline: 0;
  border-radius: 4px;
  font-weight: bold;
`;

export default buttonStyles;
