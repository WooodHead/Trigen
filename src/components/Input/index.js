/**
  * A simple styled controlled input element
  *
  * @namespace Input
  */

import React from 'react';
import styled from 'styled-components';


const StyledInput = styled.input`
  display: block;
  margin: 0 0 10px 0;
  padding: 5px 10px;
  border: 1px solid ${props => props.error ? props.theme.warning : props.theme.outlines};
  border-radius: 1px;
  background: ${props => props.theme.noneWhite};
  color: ${props => props.theme.text};
  outline: 0;

  &:placeholder-shown,
  &:-ms-input-placeholder {
    color: ${props => props.theme.outlines};
  }
`;

const Input = props => {
  return (
    <StyledInput {...props} />
  );
};

export default Input;
