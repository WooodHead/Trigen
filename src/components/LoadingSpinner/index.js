import React from 'react';

import spinnerRing from 'img/spinner_ring.gif';

import Wrapper from './Wrapper';
import SpinnerImg from './SpinnerImg';


const LoadingSpinner = () => {
  return (
    <Wrapper>
      <SpinnerImg src={spinnerRing} alt="loading"/>
    </Wrapper>
  );
};

export default LoadingSpinner;
