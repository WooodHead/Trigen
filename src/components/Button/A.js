/**
  * Renders an Anchor with button styles
  *
  * @namespace A
  * @memberOf module:Button
  */

import styled from 'styled-components';

import buttonStyles from './buttonStyles';


const A = styled.a`${buttonStyles}`;

export default A;
