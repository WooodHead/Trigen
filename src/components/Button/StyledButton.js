/**
  * Creates a button with styles applied
  *
  * @namespace StyledButton
  * @memberOf module:Button
  */

import styled from 'styled-components';

import buttonStyles from './buttonStyles';


const StyledButton = styled.button`${buttonStyles}`;

export default StyledButton;
