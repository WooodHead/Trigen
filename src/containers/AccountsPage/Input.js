/**
  * A simple restyled Input component
  */

import styled from 'styled-components';

import Input from 'components/Input';

export default styled(Input)`
  display: inline-block;
  margin-right: 10px;
`;
