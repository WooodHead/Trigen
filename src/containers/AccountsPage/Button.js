/**
  * A restyled Button component
  */

import styled from 'styled-components';

import Button from 'components/Button';


export default styled(Button)`
  display: inline-block;
  margin: 5px;
`;
