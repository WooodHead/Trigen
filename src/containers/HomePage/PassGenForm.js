import React, { PropTypes, Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import shortid from 'shortid';

import { fetchAccounts } from 'containers/AccountsPage/ducks/actions';
import { getAccounts, getAccountListStatus } from 'containers/AccountsPage/ducks/selectors';
import { getUserID, getAuthToken } from 'containers/Auth/ducks/selectors';

import Input from 'components/Input';
import Button from 'components/Button';
import LoadingSpinner from 'components/LoadingSpinner';


export class PassGenForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      suggestions: [],
      domain: '',
      username: '',
    };
  }

  componentDidMount() {
    if(!this.props.accountListStatus.fetched) {
      this.props.fetchAccounts(this.props.userID, this.props.authToken, shortid.generate);
    }
  }

  // Field validation
  validate = () => {
    let errors = [];

    if(this.state.domain.length === 0) {
      errors = errors.concat('Domain is required!');
    }

    if(this.state.username.length === 0) {
      errors = errors.concat('Username is required!');
    }

    return errors;
  }

  // Submit handling
  handleSubmit = (e) => {
    e.preventDefault();
    const validationErrors = this.validate();

    if(validationErrors.length !== 0) {
      // TODO
    }
  }

  // Autosuggest input
  getSuggestions = (value) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;

    return (
      inputLength === 0
        ? []
        : this.props.accounts.filter(
          account => {
            const val = account.domain.toLowerCase().includes(inputValue);
            return val;
          }
        )
    );
  };

  getSuggestionValue = suggestion => suggestion.domain;

  onDomainChange = (event, { newValue }) => {
    this.setState({
      domain: newValue,
    });
  }

  onSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  }

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: [],
    });
  }

  onSuggestionSelected = (e, { suggestion }) => {
    this.setState({
      username: suggestion.username
    });
  }

  renderSuggestion = suggestion => {
    return (
      <div>
        {`${suggestion.domain} - Username: ${suggestion.username}`}
      </div>
    );
  };

  // Username input
  onUsernameChange = e => {
    this.setState({
      username: e.target.value,
    });
  }

  clearUsername = () => {
    this.setState({ username: '' });
  }

  render() {
    const inputProps = {
      placeholder: 'Domain eingeben',
      value: this.state.domain,
      onChange: this.onDomainChange,
      type: 'search',
    };

    if(this.props.accountListStatus.isLoading) {
      return <LoadingSpinner />;
    }

    return (
      <form onSubmit={this.handleSubmit}>
        <Input
          autoSuggest
          getSuggestionValue={this.getSuggestionValue}
          highlightFirstSuggestion={true}
          inputProps={inputProps}
          onSuggestionSelected={this.onSuggestionSelected}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          renderSuggestion={this.renderSuggestion}
          suggestions={this.state.suggestions}
        />
        <Input
          onFocus={this.clearUsername}
          placeholder="Username eingeben"
          value={this.state.username}
          onChange={this.onUsernameChange}
        />
        <Button submit>Generiere Passwort</Button>
      </form>
    );
  }
}

PassGenForm.propTypes = {
  accountListStatus: PropTypes.object,
  accounts: PropTypes.array,
  authToken: PropTypes.string,
  fetchAccounts: PropTypes.func,
  userID: PropTypes.number,
};

export const mapStateToProps = state => {
  return {
    accounts: getAccounts(state),
    accountListStatus: getAccountListStatus(state),
    userID: getUserID(state),
    authToken: getAuthToken(state),
  };
};

export const mapDispatchToProps = dispatch => {
  return bindActionCreators({
    fetchAccounts,
  }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(PassGenForm);
