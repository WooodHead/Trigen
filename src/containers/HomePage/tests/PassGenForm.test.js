import React from 'react';
import { shallow } from 'enzyme';

import LoadingSpinner from 'components/LoadingSpinner';
import Input from 'components/Input';
import Button from 'components/Button';

import { PassGenForm } from '../PassGenForm';


const baseProps = {
    accountListStatus: {
      isLoading: false,
      fetched: false,
      errored: false,
    },
    accounts: [
      {
        id: 1,
        fakeID: 'asdfas',
        username: 'tester',
        domain: 'testDomain.net',
      }
    ],
    authToken: 'asdf',
    userID: 1,
};

const setUp = newProps => {
  const props = {
    ...baseProps,
    fetchAccounts: jest.fn(),
    ...newProps,
  };

  const enzymeWrapper = shallow(<PassGenForm {...props} />);

  return {
    props,
    enzymeWrapper,
  };
};

describe('<PassGenForm />', () => {
  it('should have initial component state', () => {
    const { enzymeWrapper } = setUp();

    expect(enzymeWrapper.state()).toMatchSnapshot();
  });

  it('should call fechtAccounts() if !fetched inside componentDidMount', () => {
    const { enzymeWrapper, props } = setUp();

    enzymeWrapper.instance().componentDidMount();
    expect(props.fetchAccounts).toHaveBeenCalledTimes(1);
  });

  it('should not call fetchAccounts() if already fetched', () => {
    const { enzymeWrapper, props } = setUp({accountListStatus: { fetched: true }});

    enzymeWrapper.instance().componentDidMount();
    expect(props.fetchAccounts).toHaveBeenCalledTimes(0);
  });

  describe('validate()', () => {
    it('should return errors for empyt domain and empty username', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.setState({ domain: '', username: ''});
      expect(enzymeWrapper.instance().validate()).toMatchSnapshot();
    });

    it('should return an empty error if there are no errors', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.setState({ domain: 'test@test.com', username: 'tester' });
      expect(enzymeWrapper.instance().validate()).toMatchSnapshot();
    })
  });

  describe('getSuggestions()', () => {
    it('should return array with suggestions on match', () => {
      const { enzymeWrapper } = setUp();

      expect(enzymeWrapper.instance().getSuggestions('test')).toMatchSnapshot();
    });

    it('should return empty array on no match', () => {
      const { enzymeWrapper } = setUp();

      expect(enzymeWrapper.instance().getSuggestions('uuu')).toEqual([]);
    });
  });

  describe('getSuggestionValue()', () => {
    it('should return suggestion value', () => {
      const { enzymeWrapper } = setUp();

      expect(enzymeWrapper.instance().getSuggestionValue({ username: 'don', domain: 'brrzz@brozz.net' }))
      .toBe('brrzz@brozz.net');
    });
  });

  describe('onDomainChange()', () => {
    it('should handle domain change', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.instance().onDomainChange({}, { newValue: 'test@test.net' });
      expect(enzymeWrapper.state()).toMatchSnapshot();
    });
  });

  describe('onSuggestionsFetchRequested()', () => {
    it('should handle suggestion fetching', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.instance().onSuggestionsFetchRequested({ value: 'test' });
      expect(enzymeWrapper.state()).toMatchSnapshot();
    });
  });

  describe('onSuggestionsClearRequested()', () => {
    it('should clear suggestions in component state', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.setState({suggestions: [{domain: 'hi'}]});

      expect(enzymeWrapper.state().suggestions).toEqual([{ domain: 'hi' }]);

      enzymeWrapper.instance().onSuggestionsClearRequested();

      expect(enzymeWrapper.state().suggestions).toEqual([])

    });
  });

  describe('onSuggestionSelected()', () => {
    it('should populate username with suggestion value', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.instance().onSuggestionSelected({}, { suggestion: { domain: 'cthulhu@was.here', username: 'fthang'}});

      expect(enzymeWrapper.state().username).toBe('fthang');
    });
  });

  describe('renderSuggestion()', () => {
    it('should return <div> and domain and username', () => {
      const { enzymeWrapper } = setUp();

      expect(enzymeWrapper.instance().renderSuggestion({ domain: 'howdy@bam.de', username: 'lucky luke'}))
        .toEqual(<div>howdy@bam.de - Username: lucky luke</div>);
    });
  });

  describe('onUsernameChange()', () => {
    it('should update username inside component state', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.instance().onUsernameChange({ target: { value: 'the devil'} });

      expect(enzymeWrapper.state().username).toBe('the devil');
    });
  });

  describe('clearUsername()', () => {
    it('should empty username inside component state', () => {
      const { enzymeWrapper } = setUp();

      enzymeWrapper.setState({ username: 'princess bubblegum' });

      expect(enzymeWrapper.state().username).toBe('princess bubblegum');

      enzymeWrapper.instance().clearUsername();

      expect(enzymeWrapper.state().username).toBe('');
    });
  });

  it('should render <LoadingSpinner /> on accountList loading', () => {
    const { enzymeWrapper } = setUp({ accountListStatus: {
      ...baseProps.accountListStatus, isLoading: true
    }});

    expect(enzymeWrapper.find(LoadingSpinner).exists()).toBe(true);
  });

  it('should render <form>', () => {
    const { enzymeWrapper } = setUp();

    expect(enzymeWrapper.find('form').exists()).toBe(true);
  })

  it('should render <Input/>', () => {
    const { enzymeWrapper } = setUp();

    expect(enzymeWrapper.find(Input).length).toBe(2)
  });

  it('should render <Button/>', () => {
    const { enzymeWrapper } = setUp();

    expect(enzymeWrapper.find(Button).exists()).toBe(true);
  });
});
