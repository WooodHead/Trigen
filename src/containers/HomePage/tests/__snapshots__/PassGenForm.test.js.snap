exports[`<PassGenForm /> getSuggestions() should return array with suggestions on match 1`] = `
Array [
  Object {
    "domain": "testDomain.net",
    "fakeID": "asdfas",
    "id": 1,
    "username": "tester",
  },
]
`;

exports[`<PassGenForm /> onDomainChange() should handle domain change 1`] = `
Object {
  "domain": "test@test.net",
  "suggestions": Array [],
  "username": "",
}
`;

exports[`<PassGenForm /> onSuggestionsFetchRequested() should handle suggestion fetching 1`] = `
Object {
  "domain": "",
  "suggestions": Array [
    Object {
      "domain": "testDomain.net",
      "fakeID": "asdfas",
      "id": 1,
      "username": "tester",
    },
  ],
  "username": "",
}
`;

exports[`<PassGenForm /> react-redux should map dispatch to props 1`] = `
Object {
  "fetchAccounts": [Function],
}
`;

exports[`<PassGenForm /> react-redux should map state to props 1`] = `
Object {
  "accountListStatus": Object {
    "isLoading": true,
  },
  "accounts": Array [
    Object {
      "domain": "hi.com",
      "username": "jonny",
    },
  ],
  "authToken": "abc",
  "userID": "123",
}
`;

exports[`<PassGenForm /> should have initial component state 1`] = `
Object {
  "domain": "",
  "suggestions": Array [],
  "username": "",
}
`;

exports[`<PassGenForm /> validate() should return an empty error if there are no errors 1`] = `Array []`;

exports[`<PassGenForm /> validate() should return errors for empyt domain and empty username 1`] = `
Array [
  "Domain is required!",
  "Username is required!",
]
`;
